core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -832,4 +834,24 @@ public void run() {
     public AliasFilter buildAliasFilter(ClusterState state, String index, String... expressions) {
         return indicesService.buildAliasFilter(state, index, expressions);
     }
+
+    /**
+     * This method does a very quick rewrite of the query and returns true if the query can potentially match any documents.
+     * This method can have false positives while if it returns <code>false</code> the query won't match any documents on the current
+     * shard.
+     */
+    public boolean canMatch(ShardSearchRequest request) throws IOException {
+        try (DefaultSearchContext context = createSearchContext(request, defaultSearchTimeout, null)) {
+            SearchSourceBuilder source = context.request().source();
+            if (source != null) {
+                QueryBuilder queryBuilder = source.query();
+                AggregatorFactories.Builder aggregations = source.aggregations();
+                boolean hasGlobalAggs = aggregations != null && aggregations.hasGlobalAggregationBuilder();
```

Comment:
As a first step, I'd just do `instanceof` checks for `TermsAggregationBuilder` and `(Date)HistogramAggregationBuilder`, and check the value of `minDocCount`.